File: pegasusX/apps/driver-app-android/app/src/main/java/com/pegasusx/driver/ui/navigation/HandoffPathResolver.kt
package com.pegasusx.driver.ui.navigation

import java.net.URI

/** Parsed native destination for a portal-style handoff `primary_link`. */
sealed class HandoffDestination {
    data object Home : HandoffDestination()
    data object FleetMap : HandoffDestination()
    data object ManifestList : HandoffDestination()
    data class ManifestDetail(val manifestId: String) : HandoffDestination()
    data class OrderDetail(val orderId: String) : HandoffDestination()
    data object Unresolved : HandoffDestination()
}

object HandoffPathResolver {
    fun resolve(link: String): HandoffDestination {
        val trimmed = link.trim()
        if (trimmed.isBlank()) return HandoffDestination.Unresolved

        val path = runCatching {
            if (trimmed.startsWith("/")) trimmed else URI(trimmed).path
        }.getOrDefault(trimmed).substringBefore('?').trimEnd('/')

        val segments = path.split('/').filter { it.isNotBlank() }
        if (segments.isEmpty()) return HandoffDestination.Unresolved

        return when (segments.first()) {
            "manifests" -> when {
                segments.size >= 2 -> HandoffDestination.ManifestDetail(segments[1])
                else -> HandoffDestination.ManifestList
            }
            "dispatch" -> HandoffDestination.Home
            "fleet" -> HandoffDestination.FleetMap
            "orders" -> when {
                segments.size >= 2 -> HandoffDestination.OrderDetail(segments[1])
                else -> HandoffDestination.Unresolved
            }
            else -> HandoffDestination.Unresolved
        }
    }
}

File: pegasusX/apps/driver-app-ios/driverappios/driverappios/Utilities/HandoffPathResolver.swift
import Foundation

enum HandoffDestination: Equatable {
    case home
    case fleetMap
    case manifestList
    case manifestDetail(String)
    case orderDetail(String)
    case unresolved
}

enum HandoffPathResolver {
    static func resolve(_ link: String) -> HandoffDestination {
        let trimmed = link.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return .unresolved }

        let path: String
        if trimmed.hasPrefix("/") {
            path = String(trimmed.split(separator: "?").first ?? Substring(trimmed))
        } else if let url = URL(string: trimmed) {
            path = url.path
        } else {
            path = trimmed
        }

        let segments = path.split(separator: "/").map(String.init).filter { !$0.isEmpty }
        guard let root = segments.first else { return .unresolved }

        switch root {
        case "manifests":
            if segments.count >= 2 { return .manifestDetail(segments[1]) }
            return .manifestList
        case "dispatch":
            return .home
        case "fleet":
            return .fleetMap
        case "orders":
            if segments.count >= 2 { return .orderDetail(segments[1]) }
            return .unresolved
        default:
            return .unresolved
        }
    }
}
